### PR TITLE
Baygur: Add Hello World label

### PR DESCRIPTION
**Pull request by yoavhacohen**

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,3 @@
 [submodule "Foundations"]
 	path = Foundations
-	url = https://github.com/Lightricks/Foundations.git
+	url = https://github.com/Lightricks/Foundations
```
[the remark this answers is not in the document]
Comment (yoavhacohen) [reply .gitmodules]: Done

File: Baygur/Baygur/BGRAppDelegate.mm [hunk 2]
```diff
@@ -13,7 +13,7 @@ @interface BGRAppDelegate ()
 @implementation BGRAppDelegate
 
 - (BOOL)application:(UIApplication __unused *)application
-    didFinishLaunchingWithOptions:(NSDictionary __unused *)launchOptions {
+didFinishLaunchingWithOptions:(nullable NSDictionary __unused *)launchOptions {
```
[the remark this answers is not in the document]
Comment (yoavhacohen) [reply Baygur/Baygur/BGRAppDelegate.mm]: Done

File: Baygur/Baygur/BGRAppDelegate.mm [hunk 3]
```diff
@@ -8,12 +8,24 @@
 NS_ASSUME_NONNULL_BEGIN
 
 @interface BGRAppDelegate ()
+
```
[the remark this answers is not in the document]
Comment (yoavhacohen) [reply Baygur/Baygur/BGRAppDelegate.mm]: Done

File: Baygur/Baygur/BGRAppDelegate.mm [hunk 4]
```diff
@@ -8,12 +8,24 @@
...
 @end
 
 @implementation BGRAppDelegate
 
 - (BOOL)application:(UIApplication __unused *)application
 didFinishLaunchingWithOptions:(nullable NSDictionary __unused *)launchOptions {
+  
+  BGRViewController* vc = [[BGRViewController alloc] init];
```
[the remark this answers is not in the document]
Comment (yoavhacohen) [reply Baygur/Baygur/BGRAppDelegate.mm]: Done

File: Baygur/Baygur/BGRAppDelegate.mm [hunk 5]
```diff
@@ -8,12 +8,24 @@
...
+  
+  CGRect frame = [[UIScreen mainScreen] bounds];
```
[the remark this answers is not in the document]
Comment (yoavhacohen) [reply Baygur/Baygur/BGRAppDelegate.mm]: Done

File: Baygur/Baygur/BGRViewController.mm [hunk 6]
```diff
@@ -12,7 +12,20 @@ @implementation BGRViewController
 
 - (void)viewDidLoad {
   [super viewDidLoad];
-
+  
+  // UILabel:
+  UILabel* label = [[UILabel alloc] initWithFrame:CGRectZero];
+  [label setTranslatesAutoresizingMaskIntoConstraints:NO];
```
[the remark this answers is not in the document]
Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: Done

File: Baygur/Baygur/BGRViewController.mm [hunk 7]
```diff
@@ -12,7 +12,20 @@ @implementation BGRViewController
 
 - (void)viewDidLoad {
   [super viewDidLoad];
-
+  
+  // UILabel:
+  UILabel* label = [[UILabel alloc] initWithFrame:CGRectZero];
```
[the remark this answers is not in the document]
Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: Done

File: Baygur/Baygur/BGRViewController.mm [hunk 8]
```diff
@@ -12,7 +12,20 @@ @implementation BGRViewController
...
+  [label setTranslatesAutoresizingMaskIntoConstraints:NO];
+  [label setText:@"Hello World"];
+  [self.view addSubview:label];
+  
+  // Layout:
+  UIView* container = self.view;
+  [label mas_makeConstraints:^(MASConstraintMaker *make) {
+    make.centerY.equalTo(container.mas_bottom).multipliedBy(1./3.);
+    make.centerX.equalTo(container.mas_centerX);
```
[the remark this answers is not in the document]
Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: Done

**Comment by yoavhacohen** [comment]
Round done.

File: Baygur/Baygur/BGRAppDelegate.mm [hunk 9]
```diff
@@ -14,6 +14,17 @@ @implementation BGRAppDelegate
 
 - (BOOL)application:(UIApplication __unused *)application
 didFinishLaunchingWithOptions:(nullable NSDictionary __unused *)launchOptions {
```
Comment (danafl) [inline Baygur/Baygur/BGRAppDelegate.mm]: indentation was not fixed yet ..

File: Baygur/Baygur/BGRViewController.mm [hunk 10]
```diff
@@ -6,13 +6,35 @@
 NS_ASSUME_NONNULL_BEGIN
 
 @interface BGRViewController ()
+
+@property (nonatomic) UILabel *label;
```
Comment (danafl) [inline Baygur/Baygur/BGRViewController.mm]: Please add docs to the property and classes according to the coding style guide.

since the label is created once it should be (readonly, nonatomic)

File: Baygur/Baygur/BGRViewController.mm [hunk 11]
```diff
@@ -6,13 +6,35 @@
...
+
 @end
 
 @implementation BGRViewController
 
+- (void)setupLabel {
+  UILabel *label = [[UILabel alloc] initWithFrame:CGRectZero];
+  [label setText:@"Hello World"];
```
Comment (danafl) [inline Baygur/Baygur/BGRViewController.mm]: `lable.text = @"Hello World"`

File: Baygur/Baygur/BGRViewController.mm [hunk 12]
```diff
@@ -6,13 +6,35 @@
...
+  [self.view addSubview:label];
+  self.label = label;
```
Comment (danafl) [inline Baygur/Baygur/BGRViewController.mm]: you can assign the lable immediatly in line 17 - 
`_label = [[UILabel alloc] initWithFrame:CGRectZero];`

File: Baygur/Baygur/BGRViewController.mm [hunk 13]
```diff
@@ -6,13 +6,35 @@
...
+}
+
+- (void)setupConstraints {
+  UIView *container = self.view;
```
Comment (danafl) [inline Baygur/Baygur/BGRViewController.mm]: We usually don't separate the constraints to a separate method - we just put it in the setup method of the label

File: Baygur/Baygur/BGRViewController.mm (hunk 10 above)
Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: done

File: Baygur/Baygur/BGRViewController.mm (hunk 12 above)
Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: I think it's a better practice to first prepare a value as a local variable before assigning it to a data member - for three reasons:
1. The type o the data member may be more generic than the concrete type we assign (think about 'NSMutableArray' as the value type and 'NSArray' as the type of the data member)
2. Setting the data member may cause updates (based on the not-ready-yet value) and should not happen more than once during the preparation of the value.
3. It saves a few unnecessary pointer indirections that can't be optimized by the compiler (at least in Objective C where there is no way to mark a data member as a constant)

One can claim that #2 is not relevant for constructor, #3 is neglectable and #1 only relevant in specific cases  - but I think it's still a good practice to always prepare values before assigning them to data members (for consistency with the cases when it's necessary).

Let me know what you think.

File: Baygur/Baygur/BGRViewController.mm (hunk 13 above)
Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: But constraints may involve multiple views... how can they be done individually per view?

File: Baygur/Baygur/BGRViewController.mm [hunk 14]
```diff
@@ -7,17 +7,18 @@
 
 @interface BGRViewController ()
 
-@property (nonatomic) UILabel *label;
+/// Label for static text
```
Comment (danafl) [inline Baygur/Baygur/BGRViewController.mm]: Add docs to the BGRViewController class as well please

File: Baygur/Baygur/BGRViewController.mm (hunk 12 above)
Comment (danafl) [reply Baygur/Baygur/BGRViewController.mm]: regarding your points - 
1. If we assign the object directly to a data member with a more generic type, that means that we don't care about it being of the concrete type, and therefore one can claim that it's even more correct  not to assign it to a variable with it's specific type. In cases where additional work needs to be done on the object as the specific type, we can first assign it to a var with its specific type and then to the generic data member, but only when it's needed.
2. Similar to the previous point - I think that if there is a good reason not to assign the object directly, its OK to first prepare it. But in general I think it's less readable and therefore should be done only when needed.
3. I agree that it neglectable :)

To sum up, I think that this approach has advantages only in specific cases, so it should be used when they are relevant.

File: Baygur/Baygur/BGRViewController.mm (hunk 13 above)
Comment (danafl) [reply Baygur/Baygur/BGRViewController.mm]: We usually setup the view in an order of dependency - if view A 's constraints depend on view B, View B will be setup before view A. However there are cases where the constrains are more complicated and we do them in a separate method like you did.

In general, this is more of a convention, and if you prefer to do it separately I think it's OK.

**Comment by danafl** [comment]
@yoavhacohen Please make sure to write a response on every comment I write in the review, in order to let me know that you fixed it or have an issue with it.

Also, don't forget to write a comment when you are done pushing the changes.

Please fix and squash.

File: Baygur/Baygur/BGRViewController.mm (hunk 12 above)
Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: OK, done

File: Baygur/Baygur/BGRViewController.mm (hunk 11 above)
Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: done

File: Baygur/Baygur/BGRAppDelegate.mm (hunk 9 above)
Comment (yoavhacohen) [reply Baygur/Baygur/BGRAppDelegate.mm]: done

File: Baygur/Baygur/BGRViewController.mm (hunk 13 above)
Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: OK, I'm keeping it in a a separate method

File: Baygur/Baygur/BGRViewController.mm (hunk 14 above)
Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: done

**Comment by yoavhacohen** [comment]
fixed and squashed

File: Baygur/Baygur/BGRAppDelegate.mm (hunk 3 above)
Comment (danafl) [inline Baygur/Baygur/BGRAppDelegate.mm]: This was not reverted... (remove new line)

File: Baygur/Baygur/BGRViewController.mm [hunk 15]
```diff
@@ -6,13 +6,36 @@
 NS_ASSUME_NONNULL_BEGIN
 
 @interface BGRViewController ()
+
+/// Label for static text
+@property (readonly, nonatomic) UILabel *label;
+
 @end
 
+/// A main container for all content in the app
```
Comment (danafl) [inline Baygur/Baygur/BGRViewController.mm]: This should be on the interface , rather than on the implementation.
Also, lease change this to be more specific, something like:
`/// View controller presenting a "Hello world" label.`

File: Baygur/Baygur/BGRViewController.mm [hunk 16]
```diff
@@ -6,13 +6,36 @@
...
 @implementation BGRViewController
 
+- (void)setupLabel {
+  _label = [[UILabel alloc] initWithFrame:CGRectZero];
+  _label.text = @"Hello World";
```
Comment (danafl) [inline Baygur/Baygur/BGRViewController.mm]: after you assign the object to the property, you can access it using "self.label". "_label" is needed because we assign the object to a readonly property.

File: Baygur/Baygur/BGRViewController.mm [hunk 17]
```diff
@@ -6,13 +6,36 @@
...
+  [self.view addSubview:_label];
```
Comment (danafl) [inline Baygur/Baygur/BGRViewController.mm]: same here - self.label

File: Baygur/Baygur/BGRAppDelegate.mm (hunk 3 above)
Comment (yoavhacohen) [reply Baygur/Baygur/BGRAppDelegate.mm]: done

File: Baygur/Baygur/BGRViewController.mm (hunk 15 above)
Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: done

File: Baygur/Baygur/BGRViewController.mm (hunk 17 above)
Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: done

File: Baygur/Baygur/BGRViewController.mm (hunk 16 above)
Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: done

**Comment by yoavhacohen** [comment]
Fixed and squashed

File: Baygur/Baygur/BGRViewController.mm [hunk 18]
```diff
@@ -5,14 +5,37 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+/// View controller presenting a "Hello world" label.
```
Comment (danafl) [inline Baygur/Baygur/BGRViewController.mm]: Sorry for the mixup - I meant the interface in the .h file

Comment (yoavhacohen) [reply Baygur/Baygur/BGRViewController.mm]: done

**Comment by yoavhacohen** [comment]
done
